File: 13/sol.py
import re
from sys import stdin

def parse(line):
    regex = r'(\d+): (\d+)\n'
    rng, depth = re.match(regex, line).groups()
    return int(rng), int(depth)

def build(lines):
    d = {rng: depth for rng, depth in lines}
    return d

def severity(d, delay):
    l = [k * v for k, v in d.items() if ((k+delay) % (v + v - 2)) == 0]
    return l

def part1(d):
    l = severity(d, 0)
    return sum(l)

def part2(d):
    i = 0
    while len(l := severity(d, i)) != 0:
        #print(s)
        i += 1
    return i

lines = [parse(line) for line in stdin]

d = build(lines)
#print(d)

# p1
print(part1(d))

# p2
print(part2(d))

File: 13/sol.rs
use std::io::{self, BufRead};

fn severity(v: &Vec<(usize, usize)>, delay: usize) -> Option<usize> {
    let mut s = 0;
    let mut avoids = true;
    for &(d, r) in v {
        if (d + delay) % (r + r - 2) == 0 {
            avoids = false;
            s += d * r;
        }
    }
    if avoids {
        return None;
    } else {
        return Some(s);
    }
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let v = stdin
        .lock()
        .lines()
        .map(|res_s| {
            let s = res_s.unwrap();
            let mut binding = s.split(": ");
            let range = binding.next().unwrap().parse::<usize>().unwrap();
            let depth = binding.next().unwrap().parse::<usize>().unwrap();
            return (range, depth);
        }).collect::<Vec<_>>();

    let s = severity(&v, 0).unwrap();
    println!("{}", s);

    let mut i = 0;
    while let Some(sev) = severity(&v, i) {
        i += 1;
    }
    println!("{}", i);


    Ok(())    
}


